File: hw/vaca/u3/HWObserver/Observer/src/main/java/ec/edu/espe/observer/model/Stock.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package ec.edu.espe.observer.model;

import ec.edu.espe.observer.view.IInvestor;
import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author LABS-ESPE
 */
public abstract class Stock {
    protected String symbol;
    protected double price;
    private List<IInvestor> observers = new ArrayList<>();

    public void addObserver(IInvestor investor) {
        observers.add(investor);
    }

    public void deleteObserver(IInvestor investor) {
        observers.remove(investor);
    }

    public void notifyObservers(Object arg) {
        for (IInvestor investor : observers) {
            investor.update(this, arg);
        }
    }

    public String getSymbol() {
        return symbol;
    }

    public double getPrice() {
        return price;
    }

    public void setSymbol(String symbol) {
        this.symbol = symbol;
        notifyObservers(symbol);
    }
}

File: hw/unda/U3/HW23Observer/Observer/model/stock.py
# model/stock.py

class Stock:
    def __init__(self, symbol, price):
        self.symbol = symbol
        self.price = price
        self._observers = []

    def add_observer(self, investor):
        self._observers.append(investor)

    def remove_observer(self, investor):
        self._observers.remove(investor)

    def notify_observers(self, data):
        for investor in self._observers:
            investor.update(self, data)

    def set_price(self, new_price):
        self.price = new_price
        self.notify_observers(new_price)

    def set_symbol(self, new_symbol):
        self.symbol = new_symbol
        self.notify_observers(new_symbol)
